src/platform_impl/linux/kms/event_loop.rs

```diff
@@ -0,0 +1,701 @@
+        let pending_user_events_clone = pending_user_events.clone();
+        let (user_events_sender, user_events_channel) = calloop::channel::channel();
+
+        // User events channel.
+        handle
+            .insert_source(user_events_channel, move |event, _, _| {
+                if let calloop::channel::Event::Msg(msg) = event {
+                    pending_user_events_clone.borrow_mut().push(msg);
+                }
+            })
+            .unwrap();
+
+        // An event's loop awakener to wake up for redraw events from winit's windows.
+        let (event_loop_awakener, event_loop_awakener_source) = calloop::ping::make_ping().unwrap();
+
+        let event_sink = EventSink::new();
+
+        // Handler of redraw requests.
+        handle
+            .insert_source(
+                event_loop_awakener_source,
+                move |_event, _metadata, data| {
+                    data.push(Event::RedrawRequested(window_id));
+                },
+            )
+            .unwrap();
+
+        // It is an Arc<Mutex<>> so that windows can change the cursor position
+        let cursor_arc = Arc::new(Mutex::new(PhysicalPosition::new(0.0, 0.0)));
+
+        // Our input handler
+        let input_backend: LibinputInputBackend = LibinputInputBackend::new(
+            input,
+            (disp_width.into(), disp_height.into()), // plane, fb
+            event_loop.handle(),
+            state,
+            keymap,
+            xkb_compose,
+            cursor_arc.clone(),
+        );
+
+        // When an input is received, add it to our EventSink
+        let input_loop: calloop::Dispatcher<'static, LibinputInputBackend, EventSink> =
+            calloop::Dispatcher::new(
+                input_backend,
+                move |event, _metadata, data: &mut EventSink| {
+                    data.push(event);
+                },
+            );
+
+        handle.register_dispatcher(input_loop).unwrap();
+
+        let window_target = event_loop::EventLoopWindowTarget {
+            p: platform_impl::EventLoopWindowTarget::Kms(EventLoopWindowTarget {
+                connector: con.clone(),
+                crtc: *crtc,
+                device: drm,
+                plane: p_plane,
+                cursor_arc,
+                event_loop_handle: handle,
+                event_sink,
+                event_loop_awakener,
+                _marker: PhantomData,
+            }),
+            _marker: PhantomData,
+        };
+
+        Ok(EventLoop {
+            event_loop,
+            pending_user_events,
+            user_events_sender,
+            window_target,
+        })
+    }
+
+    pub fn run<F>(mut self, callback: F) -> !
+    where
+        F: FnMut(Event<'_, T>, &event_loop::EventLoopWindowTarget<T>, &mut ControlFlow) + 'static,
+    {
+        let exit_code = self.run_return(callback);
+        std::process::exit(exit_code);
+    }
+
+    pub fn run_return<F>(&mut self, mut callback: F) -> i32
```

Comment:
Since this uses `calloop`, Wayland uses `calloop`, and the X11 backend will eventually use `calloop` as well (#2777), it might be nice to extract some of the common code here into a separate module. Not now, but eventually.